pkg/controller/proxyconfig/validation.go

```diff
@@ -0,0 +1,346 @@
+	"k8s.io/apimachinery/pkg/util/validation"
+)
+
+const (
+	// proxyProbeMaxRetries is the number of times to attempt an http GET
+	// to a readinessEndpoints endpoint.
+	proxyProbeMaxRetries = 3
+	// clusterConfigMapKey is the name of the data key containing the
+	// PEM-encoded CA trust bundle.
+	clusterConfigMapKey = "ca-bundle.crt"
+	proxyHTTPScheme     = "http"
+	proxyHTTPSScheme    = "https"
+	// noProxyWildcard is the string used to as a wildcard attached to a
+	// domain suffix in proxy.spec.noProxy to bypass proxy for httpProxy
+	// and httpsProxy.
+	noProxyWildcard = "*"
+	// certPEMBlock is the type taken from the preamble of a PEM-encoded structure.
+	certPEMBlock = "CERTIFICATE"
+)
+
+// ValidateProxyConfig ensures that proxyConfig is valid.
+func (r *ReconcileProxyConfig) ValidateProxyConfig(proxyConfig *configv1.ProxySpec) error {
+	if specHTTPProxySet(proxyConfig) {
+		scheme, err := validateURI(proxyConfig.HTTPProxy)
+		if err != nil {
+			return fmt.Errorf("invalid httpProxy URI: %v", err)
+		}
+		if scheme != proxyHTTPScheme {
+			return fmt.Errorf("httpProxy requires an %q URI scheme", proxyHTTPScheme)
+		}
+	}
+
+	var readinessCerts []*x509.Certificate
+	if specHTTPSProxySet(proxyConfig) {
+		scheme, err := validateURI(proxyConfig.HTTPSProxy)
+		if err != nil {
+			return fmt.Errorf("invalid httpsProxy URI: %v", err)
+		}
+		if scheme == proxyHTTPSScheme {
+			// A trusted CA bundle must be provided when using https
+			// between client and proxy.
+			if !specTrustedCASet(proxyConfig) {
+				return errors.New("trustedCA is required when using an https scheme with httpsProxy")
+			}
+			certBundle, err := r.validateTrustedCA(proxyConfig)
+			if err != nil {
+				return fmt.Errorf("failed validating TrustedCA %q: %v", proxyConfig.TrustedCA.Name, err)
+			}
+			copy(certBundle, readinessCerts[0:])
+		}
+	}
+
+	if specNoProxySet(proxyConfig) {
+		if proxyConfig.NoProxy != noProxyWildcard {
+			for _, v := range strings.Split(proxyConfig.NoProxy, ",") {
+				v = strings.TrimSpace(v)
+				errDomain := validateDomainName(v, false)
+				_, _, errCIDR := net.ParseCIDR(v)
+				if errDomain != nil && errCIDR != nil {
+					return fmt.Errorf("invalid noProxy: %v", v)
+				}
+			}
+		}
+	}
+
+	if specReadinessEndpoints(proxyConfig) {
+		for _, endpoint := range proxyConfig.ReadinessEndpoints {
+			scheme, err := validateURI(endpoint)
+			if err != nil {
+				return fmt.Errorf("invalid URI for endpoint %s: %v", endpoint, err)
+			}
+			switch {
+			case scheme == proxyHTTPScheme:
+				if err := validateHTTPReadinessEndpoint(endpoint); err != nil {
+					return fmt.Errorf("readinessEndpoint probe failed for endpoint %s", endpoint)
+				}
+			case scheme == proxyHTTPSScheme:
+				if !specTrustedCASet(proxyConfig) {
+					return fmt.Errorf("readinessEndpoint with an %q scheme requires trustedCA to be set", proxyHTTPSScheme)
+				}
+				if err := validateHTTPSReadinessEndpoint(readinessCerts, endpoint); err != nil {
```

Comment:
does readinessCerts include the system certs we're going to be augmenting with the user-provided certs?  i think it should.

(and if the user provided no certs, then it would just be the base system certs)

Reply:
I am renaming `readinessCerts` to `trustedCAcerts`. This var contains the list of validated certs from `trustedCA`. The validation ensures the configmap reference is the merged user/system bundle name and the list contains pem encoded certs.

Reply:
> and if the user provided no certs, then it would just be the base system certs

I need to update the cert publishing controller to meet this requirement.

Reply:
https://github.com/openshift/cluster-network-operator/pull/271 has been updated to a) build off this PR and b) to support "if the user provided no certs, then it would just be the base system certs" requirement.

Reply:
> does readinessCerts include the system certs we're going to be augmenting with the user-provided certs?

@bparees yes, if add'l certs are provided from the installer or creating configmap `openshift-config-managed/user-ca-bundle`. Otherwise it's a list of certs from the system trust bundle. __Note:__ `readinessCerts ` has been renamed to `certBundle` here: https://github.com/openshift/cluster-network-operator/pull/271/files#diff-1502a5b3db6bd911880e1d0d9346b561R74

Reply:
sounds good.